### DISCUSSION TITLE
app error design

### DISCUSSION THREAD

**Discussion by bellaabdelouahab** [ISIL-ESTE/Student-Workflow-Organizer#57] [Q&A] [answered]
do we really need the prop status in appError 
example
```js
new AppError(404, 'fail', 'No document found with that id');
```

and also in some responses
```js
res.status(201).json({
      status: 'success',
      data: {
        doc,
      },
    });
```

**Comment by bellaabdelouahab** [comment]
@muttaqin1

**Comment by ghost** [comment]
The statusCode is enough to describe the response type .Thats why i think we don't need the status property.
we can use this appError class to create some api specific errors by extending the class . such as : new ForbiddenError() , new BadRequestError()  
what do you think?
@bellaabdelouahab

**Reply by bellaabdelouahab** [reply]
aid Mubarak first hhh, 
i hope you re ready to continue where we left off.
about the error class what you mean by forbidden? is it like a custom msg or what? why would you extend the app error?
and btw there are many changes in the pr that haven't approved yet 
@Khalid1G please review and approve those changes

**Comment by ghost** [comment]
Eid Mubarak to you too, my friend. I haven't returned home yet. I was so busy in those days.
Furthermore, I was a **Neovim** user, but recently I have switched to **VS Code,** it will take some time to be comfortable with it.
You create some issues and assign me into it. When I will return home, I will start resolving those issues.

**So extending error class will be very helpful.**

Now we throw error like this:
```js
//bad request error
throw new AppError(400,'fail', 'Bad request');
//forbidden error
throw new AppError(403,'fail', 'Forbidden!');
```
But if we create specific errors such as:
```js
//extend the base(AppError) class
//bad Request Error
class BadRequestError extends AppError {
constructor(message = "Bad request"){
super(400, 'fail', message)
} 
}
//Forbidden Error
class ForbiddenError extends AppError {
constructor(message = "Forbidden"){
super(403 'fail', message)
} 
}
```

//now we can throw error like this
```js
if(some-logic){
throw new ForbiddenError('you are not allowed to perform this action');
}else if(some-logic){
throw new BadRequestError('data is required');
}else{
return 0;
}
```
@bellaabdelouahab

**Reply by ghost** [reply]
@bellaabdelouahab

**Reply by bellaabdelouahab** [reply]
Ah sorry i wrote a reply but i forgot to send it because i went looking for resources, anyway, why would you want to make them classes if the only thing that changes is the message?
@muttaqin1

**Reply by ghost** [reply]
**Clear and descriptive ** :By creating specific error classes, we can provide more precise and meaningful error messages. This helps in understanding the cause of the error quickly and accurately.
**Code readability**: Specific error classes make the code more readable and self-explanatory. When developers encounter these custom error classes, they immediately know the type of error that occurred.

We can do this but this is not that important.
@bellaabdelouahab

**Reply by bellaabdelouahab** [reply, marked as answer]
Yeah sure let's do it,

**Reply by bellaabdelouahab** [reply]
@muttaqin1

**Reply by ghost** [reply]
> Yeah sure let's do it, 

Lets hear what @Khalid1G thinks about this.  
@bellaabdelouahab

**Reply by bellaabdelouahab** [reply]
He's not here for the moment. @muttaqin1

**Reply by ghost** [reply]
He can reject this because this is not that important and we have to refactor our entire controllers this can introduce new bugs.
So we have to wait for his approval.
@bellaabdelouahab

**Comment by bellaabdelouahab** [comment]
the issue is open for more than a week and adding app error follows best practice  @muttaqin1, but anyway you can wait
